crates/sui-types/src/deny_list.rs

```diff
@@ -56,3 +60,64 @@ pub struct PerTypeDenyList {
     // Table<String, VecSet<address>>
     pub denied_addresses: Table,
 }
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct CoinDenyCap {
+    pub id: UID,
+}
+
+impl DenyList {
+    pub fn check_coin_deny_list(
+        address: SuiAddress,
+        coin_types: BTreeSet<String>,
+        object_store: &dyn ObjectStore,
+    ) -> UserInputResult {
+        Self::check_deny_list(DENY_LIST_COIN_TYPE_INDEX, address, coin_types, object_store)
+    }
+
+    fn check_deny_list(
+        type_index: usize,
+        address: SuiAddress,
+        coin_types: BTreeSet<String>,
+        object_store: &dyn ObjectStore,
+    ) -> UserInputResult {
+        let deny_list_object = match object_store.get_object(&SUI_DENY_LIST_OBJECT_ID) {
+            Ok(Some(obj)) => obj,
+            _ => {
+                return Ok(());
+            }
+        };
+        // Unwrap safe because the deny list object is created by the system.
+        let all_deny_list: DenyList = deny_list_object.to_rust().unwrap();
+        let deny_list = &all_deny_list.per_type_deny_lists[type_index];
```

Comment:
why is this access safe? what guarantees that `type_index` is present?

Reply:
When we create this object in Move it contains one entry which is for Coin. So this is guaranteed by Move.

Reply:
so `type_index` always has the value corresponding to Coin?

Reply:
ok i see now that it is just a constant that is passed in. i would prefer to just remove the parameter from the function and use the constant here to make it clearer that we only support one type, but it is not a big deal.

Reply:
Yeah the idea is that we may add more types in the future.